pkg/webhook/v1alpha2/pod/inject_webhook.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pod
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission/types"
+
+	v1 "k8s.io/api/core/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+
+	trialsv1alpha2 "github.com/kubeflow/katib/pkg/api/operators/apis/trial/v1alpha2"
+	katibmanagerv1alpha2 "github.com/kubeflow/katib/pkg/common/v1alpha2"
+)
+
+const (
+	// JobNameLabel represents the label key for the job name, the value is job name
+	JobNameLabel = "job-name"
+	// JobRoleLabel represents the label key for the job role, e.g. the value is master
+	JobRoleLabel = "job-role"
+)
+
+// For debug
+var log = logf.Log.WithName("injector-webhook")
+
+// sidecarInjector that inject metrics collect sidecar into master pod
+type sidecarInjector struct {
+	client         client.Client
+	decoder        types.Decoder
+	managerService string
+}
+
+var _ admission.Handler = &sidecarInjector{}
+
+func (s *sidecarInjector) Handle(ctx context.Context, req types.Request) types.Response {
+	pod := &v1.Pod{}
+	err := s.decoder.Decode(req, pod)
+	if err != nil {
+		return admission.ErrorResponse(http.StatusBadRequest, err)
+	}
+
+	// Check whether the pod need to be mutated
+	if !s.MutationRequired(pod) {
+		return admission.ValidationResponse(true, "")
+	}
+
+	// Get the namespace from req since the namespace in the pod is empty.
+	namespace := req.AdmissionRequest.Namespace
+	// Do mutation
+	mutatedPod, err := s.Mutate(pod, namespace)
```

Comment:
namespace can be get from pod object, so I think we can just use one argument instead, right?

Reply:
According to our experiment, we fail to get the namespace from pod.